android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java

```
@@ -26,76 +30,257 @@
+
+        return Single.zip(
+                Single.just(uploadFile),
+                sessionSingle,
+                statusSingle,
+                this::processUploadForValidationStatus
+        ).flatMapCompletable(i -> i);
     }
 
+    /**
+     * Performs the next upload step, according to the UploadValidationStatus. The normal
+     * transition of UploadValidationStatus is from REQUESTED to VALIDATION_IN_PROGRESS to
+     * REQUESTED.
+     *
+     * @param uploadFile             file being uploaded
+     * @param uploadSession          pre-signed upload session
+     * @param uploadValidationStatus upload validation status from Bridge
+     * @return completion of next upload step
+     */
     @NonNull
-    Single<UploadValidationStatus> uploadToS3(UploadFile uploadFile, UploadSession session) {
+    Completable processUploadForValidationStatus(@NonNull UploadFile uploadFile,
+                                                 @NonNull UploadSession uploadSession,
+                                                 @NonNull UploadValidationStatus uploadValidationStatus) {
+        checkNotNull(uploadFile, "uploadFile cannot be null");
+        checkNotNull(uploadSession, "uploadSession cannot be null");
+        checkNotNull(uploadValidationStatus, "uploadValidationSession cannot be null");
+
+        switch (uploadValidationStatus.getStatus()) {
+            case REQUESTED:
+                return uploadToS3(uploadFile, uploadSession);
+            case SUCCEEDED:
+            case DUPLICATE:
+                return dequeueUpload(uploadFile.filename);
+            case VALIDATION_IN_PROGRESS:
+                LOG.debug("Validation in progress for filename: " + uploadFile.filename +
+                        ", uploadId" + uploadSession.getId());
+                break;
+            case VALIDATION_FAILED:
+                LOG.debug("Validation failed for filename: " + uploadFile.filename +
```

Should this dequeueUpload as well?  Or will this fail silently and it will be attempted again?

---

Never mind, the comment at the end answer my questions, nice!